io.openems.edge.controller.ess.cycle/src/io/openems/edge/controller/ess/cycle/statemachine/Context.java

```diff
@@ -1,74 +1,103 @@
-
 	protected final Config config;
 	protected final ManagedSymmetricEss ess;
-	protected final int maxChargePower;
-	protected final int maxDischargePower;
+	protected final int allowedChargePower;
+	protected final int allowedDischargePower;
 	protected final ComponentManager componentManager;
+	protected final LocalDateTime parsedStartTime;
 	protected final State previousState;
-	private final LocalDateTime lastStateChange;
 
 	public Context(ControllerEssCycle parent, Config config, ComponentManager componentManager, ManagedSymmetricEss ess,
-			int maxChargePower, int maxDischargePower, State previousState, LocalDateTime lastStateChange) {
+			int allowedChargePower, int allowedDischargePower, LocalDateTime parsedStartTime, State previousState) {
 		super(parent);
 		this.config = config;
 		this.componentManager = componentManager;
 		this.ess = ess;
-		this.maxChargePower = maxChargePower;
-		this.maxDischargePower = maxDischargePower;
+		this.allowedChargePower = allowedChargePower;
+		this.allowedDischargePower = allowedDischargePower;
+		this.parsedStartTime = parsedStartTime;
 		this.previousState = previousState;
-		this.lastStateChange = lastStateChange;
 	}
 
 	/**
-	 * Helper for a state change. If awaiting hysteresis time exceeded switches from
-	 * currentState to NextState.
-	 *
-	 * @param currentState Used to output better log.
-	 * @param nextState    state which will be switched to.
-	 * @return boolean true if change state is allowed
+	 * Helper for a state change.
+	 * 
+	 * @param nextState which will be switched to.
+	 * @return {@link State#WAIT_FOR_STATE_CHANGE}.
 	 */
-	protected boolean waitForChangeState(State currentState, State nextState) {
-		if (LocalDateTime.now(this.componentManager.getClock()).minus(Duration.ofMinutes(this.config.standbyTime()))
-				.isAfter(this.lastStateChange)) {
-			return true;
-		}
-		this.logInfo(this.log, "Awaiting hysteresis for changing from [" + currentState + "] to [" + nextState + "]");
-		return false;
+	protected State changeToNextState(State nextState) {
+		this.getParent().setNextState(nextState);
+		return State.WAIT_FOR_STATE_CHANGE;
 	}
 
-	protected int getChargePower() {
-		// get max charge/discharge power
-		var power = Math.max(this.maxChargePower, this.config.power() * -1);
-		if (this.ess instanceof HybridEss) {
-			// modify target power by DC production for HybridEss
-			var ess = (HybridEss) this.ess;
-			power -= ess.getActivePower().orElse(0) - ess.getDcDischargePower().orElse(0);
+	/**
+	 * Gets the required AC power set-point for AC- or Hybrid-ESS.
+	 * 
+	 * @param ess           the {@link ManagedSymmetricEss}; checked for
+	 *                      {@link HybridEss}
+	 * @param hybridEssMode the {@link HybridEssMode}
+	 * @param power         the configured target power
+	 * @return the AC power set-point
+	 */
+	protected Integer getAcPower(ManagedSymmetricEss ess, HybridEssMode hybridEssMode, int power) {
```

Comment:
why did you choose to return Integer instead of int?

Reply:
null check is done in PowerConstraint apply method, actually we can return it as "int" but apply method was taking a Integer parameter, thats why I wanted to keep it same.